builder/dispatchers.go

```diff
@@ -556,3 +613,43 @@ func volume(b *builder, args []string, attributes map[string]bool, original stri
 	}
 	return nil
 }
+
+// ARG name=[value]
+//
+// Adds the variable foo to the trusted list of variables that can be passed
+// to builder using the --build-arg flag for expansion/subsitution or passing to 'run'.
+// Dockerfile author may optionally set a default value of this variable.
+func arg(b *builder, args []string, attributes map[string]bool, original string) error {
+	if len(args) != 1 {
+		return fmt.Errorf("ARG requires exactly one argument")
+	}
+
+	var (
+		name           string
+		value          string
+		isValuePresent bool
+	)
+
+	arg := args[0]
+	// 'arg' can just be a name or name-value pair
+	if strings.Contains(arg, "=") {
+		parts := strings.SplitN(arg, "=", 2)
+		name = parts[0]
+		value = parts[1]
+		isValuePresent = true
+	} else {
+		name = arg
+		isValuePresent = false
+	}
+	// add the arg to allowed list of build-time args from this step on.
+	b.allowedBuildArgs[name] = true
+
+	// If there is a default value associated with this arg then add it to the
+	// b.buildArgs if one is not already passed to the builder. The args passed
+	// to builder override the defaut value of 'arg'.
+	if _, ok := b.buildArgs[name]; !ok && isValuePresent {
```

Comment:
I think most people will think of `ARG key=""` and `ARG key=` as the same thing. 

However, you may have come up with a reason for why we need to support `ARG key` w/o a "=".  IOW, it means only include it in the RUN environment if it was specified on the cmd line.  While `ARG key=` and (`ARG key=""`) means it will always be included in the RUN env because it has a default value of "".

But it all depends on whether the idea of conditionally including it in the RUN env is desirable or not.  The purist in me knows there's a big difference between "no var" and "var set to empty string", but if most people do something like `if $foo = "" ...` then it doesn't really matter.   But changing this decision later will probably break people so we need to choose wisely.  :-)   In the end, I think I'm leaning towards "no value (explicit or default) means no env var" because that would make it consistent with the implicit ARGs (e.g. http_proxy), which should not appear in the RUN env since they won't have a default value.